File: python/백준/그래프/다이젝스트라/1238 파티.py
import sys
import heapq

input = sys.stdin.readline

N, M, X = map(int, input().split())

INF = int(1e9)
graph = [[] for _ in range(N + 1)]
reverse_graph = [[] for _ in range(N + 1)]

for _ in range(M):
    a, b, cost = map(int, input().split())
    graph[a].append((b, cost))
    reverse_graph[b].append((a, cost))  # 방향 뒤집기

def dijkstra(start, graph):
    distance = [INF] * (N + 1)
    
    queue = []
    heapq.heappush(queue, (0, start))
    distance[start] = 0
    
    while queue:
        dist, now = heapq.heappop(queue)
        
        if dist > distance[now]:
            continue
        
        for next_node, next_dist in graph[now]:
            cost = dist + next_dist
            
            if cost < distance[next_node]:
                distance[next_node] = cost
                heapq.heappush(queue, (cost, next_node))
                
    return distance

# X → i
dist_from_X = dijkstra(X, graph)

# i → X (뒤집어서 X → i)
dist_to_X = dijkstra(X, reverse_graph)

ans = 0
for i in range(1, N + 1):
    ans = max(ans, dist_to_X[i] + dist_from_X[i])

print(ans)

File: cpp/백준/그래프/다이젝스트라/1238 파티.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>
#include <algorithm>
using namespace std;

const int INF = 1e9;

vector<int> dijkstra(int start, vector<vector<pair<int, int>>>& graph, int N) {
    vector<int> distance(N + 1, INF);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> queue;
    queue.push({0, start});
    distance[start] = 0;

    while (!queue.empty()) {
        auto [dist, now] = queue.top();
        queue.pop();

        if (dist > distance[now]) continue;

        for (auto [next_node, next_dist] : graph[now]) {
            int cost = dist + next_dist;
            if (cost < distance[next_node]) {
                distance[next_node] = cost;
                queue.push({cost, next_node});
            }
        }
    }
    return distance;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M, X;
    cin >> N >> M >> X;

    vector<vector<pair<int, int>>> graph(N + 1);
    vector<vector<pair<int, int>>> reverse_graph(N + 1);

    for (int i = 0; i < M; ++i) {
        int a, b, cost;
        cin >> a >> b >> cost;
        graph[a].push_back({b, cost});
        reverse_graph[b].push_back({a, cost});
    }

    vector<int> dist_from_X = dijkstra(X, graph, N);
    vector<int> dist_to_X = dijkstra(X, reverse_graph, N);

    int ans = 0;
    for (int i = 1; i <= N; ++i) {
        ans = max(ans, dist_to_X[i] + dist_from_X[i]);
    }

    cout << ans << "\n";
    return 0;
}
